File: javascript/1608.Special-Array-With-X-Elements-Greater-Than-or-Equal-X.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var specialArray = function (nums) {
  nums.sort((a, b) => a - b);

  let i = 0;
  let prev = -1;
  let total_right = nums.length;

  while (i < nums.length) {
    if (
      nums[i] === total_right ||
      (prev < total_right && total_right < nums[i])
    ) {
      return total_right;
    }

    while (i + 1 < nums.length && nums[i] === nums[i + 1]) {
      i += 1;
    }

    prev = nums[i];
    i += 1;
    total_right = nums.length - i;
  }

  return -1;
};

File: python/1608.Special-Array-With-X-Elements-Greater-Than-or-Equal-X.py
class Solution:
    def specialArray(self, nums: List[int]) -> int:
        nums.sort()
        i, prev = 0, -1
        total_right = len(nums)

        while i < len(nums):
            if nums[i] == total_right or prev < total_right < nums[i]:
                return total_right
            
            while i + 1 < len(nums) and nums[i] == nums[i + 1]:
                i += 1

            prev = nums[i]
            i += 1
            total_right = len(nums) - i

        return -1